src/shared/vscode/commands2.ts

```diff
@@ -367,42 +368,143 @@ interface CommandInfo<T extends Callback> {
+        getLogger().debug(`commands: skipped telemetry for "${id.id}"`)
 
         return undefined
     }
 
     // Throttling occurs regardless of whether or not the instrumenter is invoked
     const span = telemetryName ? telemetry[telemetryName] : telemetry.vscode_executeCommand
     const debounceCount = info?.debounceCount !== 0 ? info?.debounceCount : undefined
-    telemetryInfo.set(id, { startTime: currentTime, debounceCount: 0 })
+    TelemetryDebounceInfo.instance.set(id, { startTime: currentTime, debounceCount: 0 })
 
     return <T extends Callback>(fn: T, ...args: Parameters<T>) =>
         span.run(span => {
-            ;(span as Metric<VscodeExecuteCommand>).record({ command: id, debounceCount })
+            ;(span as Metric<VscodeExecuteCommand>).record({
+                command: id.id,
+                debounceCount,
+                source: BaseCommandSource.findSource(args),
+            })
 
             return fn(...args)
         })
 }
 
+/**
+ * Adding an instance of this to the execution args of your {@link RegisteredCommand}
+ * will set the "source" attribute in the automatically emitted `vscode_executeCommand` metric.
+ */
+export class BaseCommandSource {
+    constructor(readonly source: string) {}
+
+    /**
+     * Returns the `source` of a {@link BaseCommandSource} if found
+     * in the given args.
+     */
+    static findSource(args: any[]): string | undefined {
+        const source = args.find(arg => arg instanceof BaseCommandSource)
+        return source ? source.source : undefined
+    }
+
+    toString() {
+        return `{ source: "${this.source}" }`
+    }
+}
+
+/**
+ * A way to use less characters to create a {@link BaseCommandSource} instance
+ */
+export function source(source: string) {
+    return new BaseCommandSource(source)
+}
+
+/**
+ * Sets the "source" attribute for the metric `vscode_executeCommand`.
+ *
+ * - This must be run in the callback of a {@link RegisteredCommand}.
+ */
+export function setTelemetrySource(source: BaseCommandSource) {
+    /**
+     * HACK: In the current implementation, we already record
+     * the "source" attribute in {@link getInstrumenter}() before
+     * running the callback of a {@link RegisteredCommand} by
+     * checking the callback args for a {@link BaseCommandSource}.
+     * So this whole function is redundant.
+     *
+     * But the reason this is done is because it is not obvious
+     * to contributors that simply having a {@link BaseCommandSource}
+     * in the callback args is enough to set the "source" attribute.
+     * So anyone reading the code will not be confused.
+     */
+    if (!(source instanceof BaseCommandSource)) {
+        // case where vscode commands can be executed without the expected args
+        return
+    }
+    telemetry.vscode_executeCommand.record({ source: source.source })
+}
+
+/**
+ * Storage of telemetry events that is used for the purposes
+ * of maintaining the count of throttled event executions.
```

Comment:
This sounds very similar to `TelemetryLogger` https://github.com/aws/aws-toolkit-vscode/blob/df40fccff2f3d55dad8bf9c581f669c7e7edeacf/src/shared/telemetry/telemetryLogger.ts#L46

If it's too risky to rethink this now, do we have a backlog item for that?

Reply:
What is the difference between the two, semantically?

Reply:
`TelemetryLogger` maintains a linear array to buffer all logs. If we wanted to we, could internally use a buffer for `TelemetryDebounceInfo` but it doesn't make sense in this case.

The point of `TelemetryDebounceInfo` class is to maintain a counter which would require us to iterate over the entire buffer whenever we want to get the count. Instead we can have a map which we can increment whenever a new metric comes in.

If I'm not seeing what you are lets create a ticket.

Reply:
It would require some refactoring/redesign. Yet it should be a backlog item.